File: src/services/microsoft-teams.ts
import axios from "axios"

// Microsoft Graph API endpoints
const GRAPH_API_ENDPOINT = "https://graph.microsoft.com/v1.0"

// Microsoft Teams integration service
class MicrosoftTeamsService {
  private accessToken: string | null = null

  // Set the access token after authentication
  setAccessToken(token: string) {
    this.accessToken = token
  }

  // Check if the service is authenticated
  isAuthenticated(): boolean {
    return !!this.accessToken
  }

  // Create a new Teams meeting and add to Outlook calendar
  async createMeeting(meetingDetails: {
    subject: string
    startDateTime: string
    endDateTime: string
    attendees: string[]
    content?: string
    location?: string
  }) {
    try {
      if (!this.accessToken) {
        throw new Error("Not authenticated with Microsoft Graph API")
      }

      // Create a Teams meeting
      const meetingResponse = await axios.post(
        `${GRAPH_API_ENDPOINT}/me/onlineMeetings`,
        {
          startDateTime: meetingDetails.startDateTime,
          endDateTime: meetingDetails.endDateTime,
          subject: meetingDetails.subject,
          participants: {
            attendees: meetingDetails.attendees.map((email) => ({
              upn: email,
            })),
          },
        },
        {
          headers: {
            Authorization: `Bearer ${this.accessToken}`,
            "Content-Type": "application/json",
          },
        }
      )

      // Create a calendar event with the Teams meeting link
      const calendarResponse = await axios.post(
        `${GRAPH_API_ENDPOINT}/me/calendar/events`,
        {
          subject: meetingDetails.subject,
          start: {
            dateTime: meetingDetails.startDateTime,
            timeZone: "UTC"
          },
          end: {
            dateTime: meetingDetails.endDateTime,
            timeZone: "UTC"
          },
          location: {
            displayName: meetingDetails.location || "Online Meeting"
          },
          body: {
            contentType: "HTML",
            content: `<p>${meetingDetails.content || ""}</p><p>Join Teams Meeting: <a href="${meetingResponse.data.joinUrl}">${meetingResponse.data.joinUrl}</a></p>`,
          },
          attendees: meetingDetails.attendees.map((email) => ({
            emailAddress: {
              address: email,
            },
            type: "required",
          })),
          isOnlineMeeting: true,
          onlineMeetingProvider: "teamsForBusiness",
          onlineMeeting: {
            joinUrl: meetingResponse.data.joinUrl,
          },
        },
        {
          headers: {
            Authorization: `Bearer ${this.accessToken}`,
            "Content-Type": "application/json",
          },
        }
      )

      return {
        meetingId: meetingResponse.data.id,
        joinUrl: meetingResponse.data.joinUrl,
        joinWebUrl: meetingResponse.data.joinWebUrl,
        calendarEventId: calendarResponse.data.id,
      }
    } catch (error) {
      console.error("Error creating Teams meeting:", error)
      throw error
    }
  }

  // Get meeting details
  async getMeeting(meetingId: string) {
    try {
      if (!this.accessToken) {
        throw new Error("Not authenticated with Microsoft Graph API")
      }

      const response = await axios.get(`${GRAPH_API_ENDPOINT}/me/onlineMeetings/${meetingId}`, {
        headers: {
          Authorization: `Bearer ${this.accessToken}`,
          "Content-Type": "application/json",
        },
      })

      return response.data
    } catch (error) {
      console.error("Error getting Teams meeting:", error)
      throw error
    }
  }

  // Get user's calendar events
  async getCalendarEvents(startDateTime?: string, endDateTime?: string) {
    try {
      if (!this.accessToken) {
        throw new Error("Not authenticated with Microsoft Graph API")
      }

      let url = `${GRAPH_API_ENDPOINT}/me/calendar/events`
      
      // Add filter for date range if provided
      if (startDateTime && endDateTime) {
        url += `?$filter=start/dateTime ge '${startDateTime}' and end/dateTime le '${endDateTime}'`
      }

      const response = await axios.get(url, {
        headers: {
          Authorization: `Bearer ${this.accessToken}`,
          "Content-Type": "application/json",
        },
      })

      return response.data.value
    } catch (error) {
      console.error("Error getting calendar events:", error)
      throw error
    }
  }
}

export const teamsService = new MicrosoftTeamsService()

File: backend/microsoft_teams.py
# import os
# import requests
# import json
# import datetime
# from flask import current_app

# class MicrosoftTeamsIntegration:
#     def __init__(self):
#         # self.client_id = os.environ.get('MS_CLIENT_ID')
#         # self.client_secret = os.environ.get('MS_CLIENT_SECRET')
#         # self.tenant_id = os.environ.get('MS_TENANT_ID')
#         self.client_id = "demo-client-id"
#         self.client_secret = "demo-client-secret"
#         self.tenant_id = "demo-tenant-id"
#         self.access_token = None
#         self.token_expires = None

#     def _get_token(self):
#         """Get an access token for Microsoft Graph API"""
#         if self.access_token and self.token_expires and datetime.datetime.now() < self.token_expires:
#             return self.access_token

#         url = f"https://login.microsoftonline.com/{self.tenant_id}/oauth2/v2.0/token"
        
#         payload = {
#             'client_id': self.client_id,
#             'scope': 'https://graph.microsoft.com/.default',
#             'client_secret': self.client_secret,
#             'grant_type': 'client_credentials'
#         }
        
#         headers = {
#             'Content-Type': 'application/x-www-form-urlencoded'
#         }
        
#         try:
#             response = requests.post(url, data=payload, headers=headers)
#             response.raise_for_status()
            
#             token_data = response.json()
#             self.access_token = token_data['access_token']
#             # Set token expiry time (subtract 5 minutes for safety)
#             self.token_expires = datetime.datetime.now() + datetime.timedelta(seconds=token_data['expires_in'] - 300)
            
#             return self.access_token
#         except Exception as e:
#             current_app.logger.error(f"Error getting Microsoft token: {str(e)}")
#             return None

#     def create_meeting(self, subject, start_time, end_time, attendees=None, content=None):
#         """
#         Create a Microsoft Teams meeting
        
#         Args:
#             subject (str): Meeting subject
#             start_time (str): ISO format start time
#             end_time (str): ISO format end time
#             attendees (list): List of attendee email addresses
#             content (str): Meeting content/description
            
#         Returns:
#             dict: Meeting details including join URL
#         """
#         token = self._get_token()
#         if not token:
#             return None
            
#         # Use the application's default user for creating meetings
#         # In a production app, you'd use delegated permissions with the actual user
#         url = "https://graph.microsoft.com/v1.0/users/scheduler@yourdomain.com/onlineMeetings"
        
#         # Format the meeting request
#         meeting_request = {
#             "startDateTime": start_time,
#             "endDateTime": end_time,
#             "subject": subject,
#             "isEntryExitAnnounced": True,
#             "allowedPresenters": "everyone",
#             "allowMeetingChat": "enabled",
#             "joinMeetingIdSettings": {
#                 "isPasscodeRequired": False
#             }
#         }
        
#         # Add meeting content if provided
#         if content:
#             meeting_request["content"] = content
            
#         headers = {
#             'Authorization': f'Bearer {token}',
#             'Content-Type': 'application/json'
#         }
        
#         try:
#             response = requests.post(url, headers=headers, data=json.dumps(meeting_request))
#             response.raise_for_status()
            
#             meeting_data = response.json()
            
#             # Return relevant meeting details
#             return {
#                 'meeting_id': meeting_data.get('id'),
#                 'join_url': meeting_data.get('joinUrl'),
#                 'join_web_url': meeting_data.get('joinWebUrl'),
#                 'subject': meeting_data.get('subject'),
#                 'start_time': meeting_data.get('startDateTime'),
#                 'end_time': meeting_data.get('endDateTime')
#             }
#         except Exception as e:
#             current_app.logger.error(f"Error creating Teams meeting: {str(e)}")
#             return None

#     def get_meeting(self, meeting_id):
#         """
#         Get details of a Microsoft Teams meeting
        
#         Args:
#             meeting_id (str): The ID of the meeting to retrieve
            
#         Returns:
#             dict: Meeting details
#         """
#         token = self._get_token()
#         if not token:
#             return None
            
#         url = f"https://graph.microsoft.com/v1.0/users/scheduler@yourdomain.com/onlineMeetings/{meeting_id}"
        
#         headers = {
#             'Authorization': f'Bearer {token}',
#             'Content-Type': 'application/json'
#         }
        
#         try:
#             response = requests.get(url, headers=headers)
#             response.raise_for_status()
            
#             meeting_data = response.json()
            
#             return {
#                 'meeting_id': meeting_data.get('id'),
#                 'join_url': meeting_data.get('joinUrl'),
#                 'join_web_url': meeting_data.get('joinWebUrl'),
#                 'subject': meeting_data.get('subject'),
#                 'start_time': meeting_data.get('startDateTime'),
#                 'end_time': meeting_data.get('endDateTime')
#             }
#         except Exception as e:
#             current_app.logger.error(f"Error getting Teams meeting: {str(e)}")
#             return None

import os
import requests
import json
import datetime
from flask import current_app

class MicrosoftTeamsIntegration:
    def __init__(self):
        # Comment out the real Microsoft credentials - will be used later
        """
        self.client_id = os.environ.get('MS_CLIENT_ID')
        self.client_secret = os.environ.get('MS_CLIENT_SECRET')
        self.tenant_id = os.environ.get('MS_TENANT_ID')
        """
        # Use mock values for demo
        self.client_id = "demo-client-id"
        self.client_secret = "demo-client-secret"
        self.tenant_id = "demo-tenant-id"
        self.access_token = None
        self.token_expires = None

    def _get_token(self):
        """Get an access token for Microsoft Graph API"""
        # For demo purposes, return a mock token
        # Comment out the actual implementation - will be used later
        """
        if self.access_token and self.token_expires and datetime.datetime.now() < self.token_expires:
            return self.access_token

        url = f"https://login.microsoftonline.com/{self.tenant_id}/oauth2/v2.0/token"
        
        payload = {
            'client_id': self.client_id,
            'scope': 'https://graph.microsoft.com/.default',
            'client_secret': self.client_secret,
            'grant_type': 'client_credentials'
        }
        
        headers = {
            'Content-Type': 'application/x-www-form-urlencoded'
        }
        
        try:
            response = requests.post(url, data=payload, headers=headers)
            response.raise_for_status()
            
            token_data = response.json()
            self.access_token = token_data['access_token']
            # Set token expiry time (subtract 5 minutes for safety)
            self.token_expires = datetime.datetime.now() + datetime.timedelta(seconds=token_data['expires_in'] - 300)
            
            return self.access_token
        except Exception as e:
            current_app.logger.error(f"Error getting Microsoft token: {str(e)}")
            return None
        """
        
        # Mock token for demo purposes
        self.access_token = "mock-access-token"
        self.token_expires = datetime.datetime.now() + datetime.timedelta(hours=1)
        return self.access_token

    def create_meeting(self, subject, start_time, end_time, attendees=None, content=None):
        """
        Create a Microsoft Teams meeting - demo implementation
        
        Args:
            subject (str): Meeting subject
            start_time (str): ISO format start time
            end_time (str): ISO format end time
            attendees (list): List of attendee email addresses
            content (str): Meeting content/description
            
        Returns:
            dict: Meeting details including join URL
        """
        # Comment out the actual implementation - will be used later
        """
        token = self._get_token()
        if not token:
            return None
            
        # Use the application's default user for creating meetings
        # In a production app, you'd use delegated permissions with the actual user
        url = "https://graph.microsoft.com/v1.0/users/scheduler@yourdomain.com/onlineMeetings"
        
        # Format the meeting request
        meeting_request = {
            "startDateTime": start_time,
            "endDateTime": end_time,
            "subject": subject,
            "isEntryExitAnnounced": True,
            "allowedPresenters": "everyone",
            "allowMeetingChat": "enabled",
            "joinMeetingIdSettings": {
                "isPasscodeRequired": False
            }
        }
        
        # Add meeting content if provided
        if content:
            meeting_request["content"] = content
            
        headers = {
            'Authorization': f'Bearer {token}',
            'Content-Type': 'application/json'
        }
        
        try:
            response = requests.post(url, headers=headers, data=json.dumps(meeting_request))
            response.raise_for_status()
            
            meeting_data = response.json()
            
            # Return relevant meeting details
            return {
                'meeting_id': meeting_data.get('id'),
                'join_url': meeting_data.get('joinUrl'),
                'join_web_url': meeting_data.get('joinWebUrl'),
                'subject': meeting_data.get('subject'),
                'start_time': meeting_data.get('startDateTime'),
                'end_time': meeting_data.get('endDateTime')
            }
        except Exception as e:
            current_app.logger.error(f"Error creating Teams meeting: {str(e)}")
            return None
        """
        
        # Mock implementation for demo purposes
        current_app.logger.info(f"Creating mock Teams meeting: {subject}")
        
        # Generate unique meeting ID
        meeting_id = f"mock-meeting-{datetime.datetime.now().timestamp()}"
        
        # Return mock meeting details
        return {
            'meeting_id': meeting_id,
            'join_url': f"https://teams.microsoft.com/l/meetup-join/{meeting_id}",
            'join_web_url': f"https://teams.microsoft.com/l/meetup-join/{meeting_id}/web",
            'subject': subject,
            'start_time': start_time,
            'end_time': end_time
        }

    def get_meeting(self, meeting_id):
        """
        Get details of a Microsoft Teams meeting - demo implementation
        
        Args:
            meeting_id (str): The ID of the meeting to retrieve
            
        Returns:
            dict: Meeting details
        """
        # Comment out actual implementation - will be used later
        """
        token = self._get_token()
        if not token:
            return None
            
        url = f"https://graph.microsoft.com/v1.0/users/scheduler@yourdomain.com/onlineMeetings/{meeting_id}"
        
        headers = {
            'Authorization': f'Bearer {token}',
            'Content-Type': 'application/json'
        }
        
        try:
            response = requests.get(url, headers=headers)
            response.raise_for_status()
            
            meeting_data = response.json()
            
            return {
                'meeting_id': meeting_data.get('id'),
                'join_url': meeting_data.get('joinUrl'),
                'join_web_url': meeting_data.get('joinWebUrl'),
                'subject': meeting_data.get('subject'),
                'start_time': meeting_data.get('startDateTime'),
                'end_time': meeting_data.get('endDateTime')
            }
        except Exception as e:
            current_app.logger.error(f"Error getting Teams meeting: {str(e)}")
            return None
        """
        
        # Return mock meeting details
        return {
            'meeting_id': meeting_id,
            'join_url': f"https://teams.microsoft.com/l/meetup-join/{meeting_id}",
            'join_web_url': f"https://teams.microsoft.com/l/meetup-join/{meeting_id}/web",
            'subject': "Mock Meeting",
            'start_time': datetime.datetime.now().isoformat(),
            'end_time': (datetime.datetime.now() + datetime.timedelta(hours=1)).isoformat()
        }
